pallas-network/src/facades.rs

```diff
@@ -81,6 +85,63 @@ impl PeerClient {
     }
 }
 
+/// Server of N2N Ouroboros
+pub struct PeerServer {
+    plexer_handle: JoinHandle<Result<(), crate::multiplexer::Error>>,
+    version: (VersionNumber, n2n::VersionData),
+    pub chainsync: chainsync::N2NServer,
+    blockfetch: blockfetch::Server,
+}
+
+impl PeerServer {
+    pub async fn accept(listener: &TcpListener, magic: u64) -> Result<Self, Error> {
+        let (bearer, _) = Bearer::accept_tcp(&listener)
+            .await
+            .map_err(Error::ConnectFailure)?;
+
+        let mut server_plexer = multiplexer::Plexer::new(bearer);
+
+        let hs_channel = server_plexer.subscribe_server(PROTOCOL_N2N_HANDSHAKE);
+        let cs_channel = server_plexer.subscribe_server(PROTOCOL_N2N_CHAIN_SYNC);
+        let bf_channel = server_plexer.subscribe_server(PROTOCOL_N2N_BLOCK_FETCH);
+
+        let mut server_hs: handshake::Server<n2n::VersionData> = handshake::Server::new(hs_channel);
+        let server_cs = chainsync::N2NServer::new(cs_channel);
+        let server_bf = blockfetch::Server::new(bf_channel);
+
+        let plexer_handle = tokio::spawn(async move { server_plexer.run().await });
+
+        let accepted_version = server_hs
+            .handshake(n2n::VersionTable::v7_and_above(magic))
+            .await
+            .map_err(Error::HandshakeProtocol)?;
+
+        if let Some(ver) = accepted_version {
+            Ok(Self {
+                plexer_handle,
+                version: ver,
+                chainsync: server_cs,
+                blockfetch: server_bf,
+            })
+        } else {
+            Err(Error::IncompatibleVersion)
```

Comment:
I think that we should explicitly abort the plexer async thread if we don't find a compatible version

Reply:
So call `self.abort()` before returning the error?